Assets/MixedRealityToolkit/Utilities/BuildAndDeploy/UwpAppxBuildTools.cs

```diff
@@ -408,6 +438,125 @@ public static void AddGazeInputCapability(IBuildInfo buildInfo)
+        /// Adds the 'Research Mode' capability to the manifest.
+        /// </summary>
+        /// <remarks>
+        /// This is only for research projects and should not be used in production.
+        /// For further information take a look at https://docs.microsoft.com/en-us/windows/mixed-reality/research-mode.
+        /// Note that this function is only public to poke a hole for testing - do not
+        /// take a dependency on this function.
+        /// </remarks>
+        public static void AddResearchModeCapability(XElement rootNode)
+        {
+            // Add rescap Namespace to package tag
+            XNamespace rescapNs = "http://schemas.microsoft.com/appx/manifest/foundation/windows10/restrictedcapabilities";
+            var rescapAttribute = rootNode.Attribute(XNamespace.Xmlns + "rescap");
+            if (rescapAttribute == null)
+            {
+                rescapAttribute = new XAttribute(XNamespace.Xmlns + "rescap", rescapNs);
+                rootNode.Add(rescapAttribute);
+            }
+
+            // Add rescap to IgnorableNamespaces
+            var ignNsAttribute = rootNode.Attribute("IgnorableNamespaces");
+            if (ignNsAttribute == null)
+            {
+                ignNsAttribute = new XAttribute("IgnorableNamespaces", "rescap");
+                rootNode.Add(ignNsAttribute);
+            }
+
+            if (!ignNsAttribute.Value.Contains("rescap"))
+            {
+                ignNsAttribute.Value += " rescap";
+            }
+
+            // If the capabilities container tag is missing, make sure it gets added.
+            var capabilitiesTag = rootNode.GetDefaultNamespace() + "Capabilities";
+            XElement capabilitiesNode = rootNode.Element(capabilitiesTag);
+            if (capabilitiesNode == null)
+            {
+                capabilitiesNode = new XElement(capabilitiesTag);
+                rootNode.Add(capabilitiesNode);
+            }
+
+            var researchModeCapability = rescapNs + "Capability";
+            XElement existingResearchModeCapability = capabilitiesNode.Elements(researchModeCapability)
+                .FirstOrDefault(element => element.Attribute("Name")?.Value == "perceptionSensorsExperimental");
+
+            // Only add the capability if isn't there already.
+            if (existingResearchModeCapability == null)
+            {
+                capabilitiesNode.Add(
+                    new XElement(researchModeCapability, new XAttribute("Name", "perceptionSensorsExperimental")));
+            }
+        }
+
+        /// <summary>
+        /// An overload of AddResearchModeCapability that will read the AppX manifest
+        /// from the build output and update the manifest file with the rescap aka
+        /// research mode capability.
+        /// </summary>
+        /// <param name="buildInfo">An IBuildInfo containing a valid OutputDirectory</param>
+        public static void AddResearchModeCapability(IBuildInfo buildInfo)
```

Comment:
Also a nit: With the change as-is, I believe that the manifest is cracked open twice. I might do a little refactoring here after - it would be worth probably just opening the manifest once (i.e. read once, write once), and making it so that we use the function above that are actually... yeah, only defined as being public for testing purposes (I think it would probably be okay do delete that portion of the comment at that time).

To be a bit clearer, it might be better to have something in UwpPlayerBuildTools.cs that instead of:

```csharp
  var uwpBuildInfo = innerBuildInfo as UwpBuildInfo;
                    Debug.Assert(uwpBuildInfo != null);
                    if (uwpBuildInfo.GazeInputCapabilityEnabled)
                    {
                        UwpAppxBuildTools.AddGazeInputCapability(uwpBuildInfo);
                    }

                    if (EditorUserBuildSettings.wsaSubtarget == WSASubtarget.HoloLens)
                    {
                        if (uwpBuildInfo.ResearchModeCapabilityEnabled)
                        {
                            UwpAppxBuildTools.AddResearchModeCapability(uwpBuildInfo);
                        }

                        if (EditorUserBuildSettings.wsaGenerateReferenceProjects && uwpBuildInfo.AllowUnsafeCode)
                        {
                            UwpAppxBuildTools.AllowUnsafeCode(uwpBuildInfo);
                        }
                    }
```

Is more like:

```csharp
  var uwpBuildInfo = innerBuildInfo as UwpBuildInfo;
                    Debug.Assert(uwpBuildInfo != null);

void UpdateManifest(UwpBuildInfo uwpBuildInfo)
{
string manifestFilePath = GetManifestFilePath(buildInfo);
            if (manifestFilePath == null)
            {
                throw new FileNotFoundException("Unable to find manifest file");
            }

            var rootElement = XElement.Load(manifestFilePath);

if (uwpBuildInfo.GazeInputCapabilityEnabled)
{
            UwpAppxBuildTools.AddGazeInputCapability(rootElement);
}

                    if (EditorUserBuildSettings.wsaSubtarget == WSASubtarget.HoloLens && uwpBuildInfo.ResearchModeCapabilityEnabled)
                    {
                            UwpAppxBuildTools.AddResearchModeCapability(uwpBuildInfo);
                    }
            rootElement.Save(manifestFilePath);
}
void UpdateAssemblyCSharp(UwpBuildInfo  uwpBuildInfo)
{
if (EditorUserBuildSettings.wsaGenerateReferenceProjects && uwpBuildInfo.AllowUnsafeCode)
                        {
                            UwpAppxBuildTools.AllowUnsafeCode(uwpBuildInfo);
                        }
}

// And then in the actual function call:
UpdateManifest(uwpBuildInfo);
UpdateAssemblyCSharp(uwpBuildInfo);

```

Obviously with better formatting than the mess of indentation that I have above

Reply:
Yeah, currently it is opened twice. I will think about it. Hm...there is already a method for updating the manifest `UpdateAppxManifest` which does a little bit more than just adding capabilities. Is it ok using that method instead of writing another `UpdateManifest` which just updates the capabilities? Or should i extract that capabilites part from `UpdateAppxManifest` and use the following method which i have already added while addressing your previous comment.
```csharp
public static void AddCapabilities(IBuildInfo buildInfo)
{
    string manifestFilePath = GetManifestFilePath(buildInfo);
    if (manifestFilePath == null)
    {
        throw new FileNotFoundException("Unable to find manifest file");
    }

    var rootElement = XElement.Load(manifestFilePath);
    var uwpBuildInfo = buildInfo as UwpBuildInfo;

    Debug.Assert(uwpBuildInfo != null);
    if (uwpBuildInfo.GazeInputCapabilityEnabled)
    {
        AddGazeInputCapability(rootElement);
    }

    if (EditorUserBuildSettings.wsaSubtarget == WSASubtarget.HoloLens)
    {
        if (uwpBuildInfo.ResearchModeCapabilityEnabled)
        {
            AddResearchModeCapability(rootElement);
        }

        if (EditorUserBuildSettings.wsaGenerateReferenceProjects && uwpBuildInfo.AllowUnsafeCode)
        {
            AllowUnsafeCode(rootElement);
        }
    }

    rootElement.Save(manifestFilePath);
}
```

Reply:
Yep, this is good! I think that there's one more thing that needs to be addressed here (which I opened a new comment for, so I'm resolving this one now)